src/tiledb/cloud/files/utils.py

```diff
@@ -264,3 +267,134 @@ def _auth_headers(cfg: configuration.Configuration) -> Mapping[str, object]:
+    storage_path = f"{storage_path.strip('/')}/{base_dir}"
+    logger.debug("Output storage path: %s" % storage_path)
+
+    access_credentials_name = access_credentials_name or acn
+
+    # Group check and/or creation
+    if not group_uri:
+        logger.debug("No group_uri provided. Choosing one...")
+        if not name.startswith("s3://"):
+            group_uri = f"tiledb://{namespace}/{base_dir}"
+        else:
+            group_uri = output_uri
+
+    group_created = False
+    if not tiledb.object_type(group_uri, ctx=tiledb.cloud.Ctx()) == "group":
+        group_namespace, group_name = utils.split_uri(group_uri)
+        groups.create(
+            name=group_name,
+            namespace=group_namespace,
+            storage_uri=storage_path,
+            credentials_name=access_credentials_name,
+        )
+        group_created = True
+        logger.debug("Group URI: '%s' created" % group_uri)
+
+    logger.info(
+        """
+        ----------------------------------------------------
+        Folder Upload Stats:
+        - Input URI: %s
+        - Output URI: %s
+        - Group URI: %s
+            - Created: %s
+        - Excluded Files: %s
+        - Flatten: %s
+        ----------------------------------------------------
+        """
+        % (input_uri, output_uri, group_uri, group_created, exclude_files, flatten)
+    )
+
+    vfs = tiledb.VFS(config=config)
+    # Create the base dir in the destination, if it does not exist.
+    if not vfs.is_dir(storage_path):
+        vfs.create_dir(storage_path)
+
+    uploaded = 0
+    dir_count = 0
+    upload_errors: Dict[str, str] = {}
+    input_ls: List[str] = vfs.ls(input_uri, recursive=True)
+    for fname in input_ls:
+        # Skip manually excluded files/folders
+        if fname in exclude_files:
+            logger.debug("- '%s' in Excluded Files. Skipping..." % fname)
+            continue
+
+        fpath = fname.split(base_dir + os.sep)[1]
+        if vfs.is_dir(fname):
+            dir_count += 1
+            # Do not create nested folder if "flatten"
+            if flatten:
+                continue
+
+            # Create any nested dir in the destination.
+            dir_path = f"{storage_path}/{fpath}"
+            if not vfs.is_dir(dir_path):
+                logger.debug("Creating sub-folder '%s'" % dir_path)
+                vfs.create_dir(dir_path)
+        else:
+            out_path, filename = os.path.split(fpath)
+            try:
+                upload_file(
+                    input_uri=fname,
+                    output_uri=f"{storage_path}/{out_path}",
+                    filename=filename,
+                    access_credentials_name=access_credentials_name,
+                )
+                uploaded += 1
+            except Exception as exc:
+                upload_errors[fname] = str(exc)
+
+    return {
+        "msg": f"Uploaded {uploaded}/{len(input_ls) - dir_count} files",
+        "errors": upload_errors,
+    }
```

Comment:
maybe make this an attrs class?

Reply:
I prefer the simplicity of dictionaries, especially as return argument.
Plus it is a serializable result.

Reply:
In both cases, I think the right solution is to implement a simple wrapper which transforms the result into something that is JSON-serializable, rather than preemptively serializing the information in the return. Doing so just means that anybody who wants to make use of it has to parse it back again rather than using a defined structure.